### PR TITLE
KNE-30: Add tests for the core submodules

### PR DESCRIPTION
**Pull request by rossnelson**
Branched off KNE-27

- kne-express.spec.js
- kne-libraries.spec.js // KNE-28
- kne-mongo.spec.js
- kne-mysql.spec.js
- kne-queue.spec.js
- kne-redis.spec.js
- kne-websocket.spec.js

To complete JSAPI module tests, JSAPI results and errors need to be mocked.
Then at least each KneJsapi prototype function should be tested. But this should be handled in a future update.

### CODE REVIEW

File: submodules/kneWebsocket.js [hunk 1]
```diff
@@ -108,7 +108,9 @@ module.exports.initMaster = function (defer, engine, log) {
                     }
                 });
 
+
                 var id = getWorkerIndex(realIp, engine.getConfig('kne.workers'));
+                if (process.env.NODE_ENV == 'testing') { id = 1; }
```
Comment (rossnelson) [inline submodules/kneWebsocket.js]: The testing ENV only creates one worker, regardless of the configuration value.

File: spec/engine.spec.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        it('throws an error on an invalid module', () => {
+            const name = 'bogus-module';
+            expect(engine.getModule.bind(engine, name)).toThrowError();
+        });
+    });
+
+    describe('#getSubmodule()', function () {
+        it('returns the specified master submodule', function () {
+            fn = engine.getSubmodule.bind(engine, 'kne-queue');
+            expect(fn() instanceof Object).toBeTruthy();
```
Comment (jgpacheco) [inline spec/engine.spec.js]: `expect(fn()).toEqual(jasmine.any(Object))` is nice.

File: spec/engine.spec.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    describe('#getConfig()', function () {
+        it('returns the specified configuration', function () {
+            const fn = engine.getConfig.bind(engine, 'test-module.test1');
+            expect(fn()).toEqual('teststring');
+        });
+
+        it('throws an error when attempting to get an undefined configuration', function () {
+            const fn = engine.getConfig.bind(engine, 'should.not.exist.fake');
+            expect(fn).toThrowError();
```
Comment (jgpacheco) [inline spec/engine.spec.js]: `fn` or `fn()`?

File: spec/engine.spec.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+        it('throws an error on an invalid module', () => {
+            const name = 'bogus-module';
+            expect(engine.getModule.bind(engine, name)).toThrowError();
+        });
+    });
+
+    describe('#getSubmodule()', function () {
+        it('returns the specified master submodule', function () {
+            fn = engine.getSubmodule.bind(engine, 'kne-queue');
```
Comment (jgpacheco) [inline spec/engine.spec.js]: You forgot the `const` in this one.

File: spec/engine.spec.js (hunk 3 above)
Comment (rossnelson) [inline spec/engine.spec.js]: When you expect a thrown error, you cannot call the function. Jasmine needs to wrap the function first so is can call and catch the thrown error.

File: spec/submodules/kne-mongo.spec.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+    describe('#init()', () => {
+        it('has a valid DB connection', () => {
+            expect(mongoose.connection.readyState).toBe(1);
+        });
+    });
+
+    describe('#loadModels()', () => {
+        it(`requires each submodule's models inside the models directory`, () => {
+            expect(mongoose.model('Test') instanceof Object).toBeTruthy();
```
Comment (jgpacheco) [inline spec/submodules/kne-mongo.spec.js]: If `mongoose.model` throws an error, maybe we can write `expect(mongoose.model('Test').not.toThrowError()`

File: spec/engine.spec.js (hunk 3 above)
Comment (jgpacheco) [inline spec/engine.spec.js]: Interesting.

**Pull request merged**